ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs

```diff
@@ -0,0 +1,215 @@
+import           Ouroboros.Consensus.Shelley.Protocol.Crypto
+
+import qualified Ouroboros.Consensus.Cardano.Block as Cardano
+import           Ouroboros.Consensus.Cardano.Node (HardCodedTransition (..),
+                     protocolInfoCardano)
+
+{-------------------------------------------------------------------------------
+  HasAnalysis
+-------------------------------------------------------------------------------}
+
+class HasAnalysis blk where
+    countTxOutputs   :: blk -> Maybe Int
+    blockHeaderSize  :: blk -> Maybe Natural
+    blockTxsSize     :: blk -> Maybe [BL.ByteString]
+    mkTopLevelConfig :: [FilePath] -- Genesis file or files.
+                     -> Bool       -- is it mainnet?
+                     -> IO (EpochInfo Identity, TopLevelConfig blk)
+
+{-------------------------------------------------------------------------------
+  ByronBlock instance
+-------------------------------------------------------------------------------}
+
+instance HasAnalysis ByronBlock where
+    countTxOutputs = whenNotEBB countTxOutputsByron
+    blockHeaderSize = whenNotEBB blockHeaderSizeByron
+    blockTxsSize = whenNotEBB blockTxsSizeByron
+    mkTopLevelConfig [configFile] onMainNet = do
+      genesisConfig <- openGenesisByron configFile onMainNet
+      let epochSlots = Genesis.configEpochSlots genesisConfig
+          epochInfo = fixedSizeEpochInfo (coerce epochSlots)
+      let config = mkByronTopLevelConfig genesisConfig
+      return (epochInfo, config)
+    mkTopLevelConfig ls _onMainNet =
+      error $
+        "a single genesis file is needed for pure Byron. Given " ++ show ls
+
+type ByronBlock = Byron.ByronBlock
+
+-- | Returns 'Nothing' for EBBs and a value for regular blocks.
+whenNotEBB :: (Chain.ABlock ByteString -> a) -> ByronBlock ->  Maybe a
+whenNotEBB k blk = case blk of
+    Byron.ByronBlock (Chain.ABOBBlock regularBlk) _ _
+      -> Just $ k regularBlk
+    _ -> Nothing
+
+countTxOutputsByron :: Chain.ABlock ByteString -> Int
+countTxOutputsByron Chain.ABlock{..} = countTxPayload bodyTxPayload
+  where
+    Chain.AHeader{..} = blockHeader
+    Chain.ABody{..}   = blockBody
+
+    countTxPayload :: Chain.ATxPayload a -> Int
+    countTxPayload = sum
+                   . map (countTx . unAnnotated . Chain.aTaTx)
+                   . Chain.aUnTxPayload
+
+    countTx :: Chain.Tx -> Int
+    countTx = length . Chain.txOutputs
+
+blockHeaderSizeByron ::  Chain.ABlock ByteString -> Natural
+blockHeaderSizeByron = Chain.headerLength . Chain.blockHeader
+
+blockTxsSizeByron :: Chain.ABlock ByteString -> [BL.ByteString]
+blockTxsSizeByron block = map (BL.fromStrict . Chain.aTaAnnotation) blockTxAuxs
+  where
+    Chain.ABlock{ blockBody } = block
+    Chain.ABody{ bodyTxPayload } = blockBody
+    Chain.ATxPayload{ aUnTxPayload = blockTxAuxs } = bodyTxPayload
+
+openGenesisByron :: FilePath -> Bool -> IO Genesis.Config
+openGenesisByron configFile onMainNet = do
+    eigenesisHash <- runExceptT $
```

Comment:
Something like:
```suggestion
    genesisHash <- either throwM return =<< runExceptT $
```

Reply:
Eventually `GenesisDataError` is not an `Exception`, so I used `error . show` instead of `throwM`